Added ability to use spot instances.

jairideout opened the pull request:
Includes new `-b/--spot_bid` option to specify maximum spot bid. Defaults to "on-demand" flat rate (mimicking `starcluster start`).

This fixes #17.

jairideout committed:
Added -b/--spot_bid option to clout script.

Still need to implement this functionality.

jairideout committed:
Added implementation for spot bidding.

Also added unit tests.

jairideout committed:
Updated documentation and examples.

gregcaporaso commented on clout/run.py:
```diff
@@ -34,18 +42,20 @@ def run_test_suites(config_f, sc_config_fp, recipients_f, email_settings_f,
+        user - the user who the tests should be run as on the cluster
+            (a string)
+        spot_bid - the maximum bid to use for worker spot instances (a float).
+            If None, "on-demand" flat rates will be used for all instances
```
I'd list the units here (i.e., `USD` if that is always what it would be - not sure what it would be for users in the asia-pacific region for example, but since you're just passing it through to StarCluster, just see how they document this) since users may not all be in the US. Also, I think it's worth a sanity check on this value (e.g., if it's more than $10 it's probably an error) just because under some very unlikely circumstances it could possibly affect how much a user pays for their compute time.

gregcaporaso commented:
looks good pending two documentation edits. you can move it over once these are merged. thanks!

gregcaporaso commented on scripts/clout:
```diff
@@ -70,12 +74,17 @@ optional_group = OptionGroup(parser, 'Optional Options')
         help='the user to run the test suites as on the remote cluster '
         '[default: %default]', default='root'),
+    make_option('-b', '--spot_bid', type='float',
+        help='the maximum bid to use for spot instances. If this option is '
```
same comment about units

jairideout commented:
Just noticed `starcluster start` has a `--force-spot-master` option. Was reading the online documentation and it said that the master node will always be flat-rate. Should have checked `start` more closely.

This will simplify the necessary changes quite a bit. I'll fix and comment when the changes are in. I'll also look into the spot bid units as StarCluster doesn't say what they are in the command-line help.

As far as sanity-checking goes, I'll set a $10 max that is on by default and provide an option to suppress this sanity check. Does this sound okay to you?

jairideout committed:
Now uses --force-spot-master.

Also updated documentation to state units for spot bids and added sanity checking (includes new option --suppress_spot_bid_check).

jairideout commented:
I have addressed the comments. @gregcaporaso, can you please review again?

gregcaporaso commented:
looks good - thanks!
